### DISCUSSION TITLE
omitEmptyFields doesn't work with %property(ENV_VAR)

### DISCUSSION THREAD

**Discussion by ziomill** [logfellow/logstash-logback-encoder#760] [Q&A] [answered]
Hello,

I'm trying to use the "**omitEmptyFields**" to print/omit an ENV variable, without success. If the ENV_VAR 'CONTAINER_NAMESPACE' is not populated, the logger prints "null" as the field's value rather than omit the field. Note that the "**omitEmptyFields**" is working when used with **%mdc** directive. This is my configuration:

trace_id

```xml <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
    <encoder class="net.logstash.logback.encoder.LoggingEventCompositeJsonEncoder">
        <providers>
            <!-- TIMESTAMP field (UTC) -->
            <timestamp>
                <fieldName>Timestamp</fieldName>
                <timeZone>UTC</timeZone>
            </timestamp>
            <!-- Include TRACE-ID field from MDC -->
            <pattern>
                <omitEmptyFields>true</omitEmptyFields>
                <pattern>
                    {"TraceId":"%mdc{trace_id}"}
                </pattern>
            </pattern>
            <!-- Include SPAN-ID field from MDC -->
            <pattern>
                <omitEmptyFields>true</omitEmptyFields>
                <pattern>
                    {"SpanId":"%mdc{span_id}"}
                </pattern>
            </pattern>
            <!-- LOG_LEVEL field -->
            <logLevel>
                <fieldName>SeverityText</fieldName>
            </logLevel>
            <!-- MESSAGE field -->
            <message>
                <fieldName>Body</fieldName>
            </message>
            <!-- RESOURCE nested object -->
            <nestedField>
                <fieldName>Resource</fieldName>
                <providers>
                    <!-- NAMESPACE field -->
                    <pattern>
                        <omitEmptyFields>true</omitEmptyFields>
                        <pattern>
                            {"container-namespace":"%property{CONTAINER_NAMESPACE}"}
                        </pattern>
                    </pattern>
                </providers>
            </nestedField>
        </providers>
    </encoder>
<appended>
```

And this is the printed log record:

```js
{
   "Timestamp":"2022-02-16T16:20:14.530Z",
   "SeverityText":"INFO",
   "Body":"gRPC Server started, listening on address: *, port: 11001",
   "Resource":{
      "container-namespace":"null"
   }
}
```

Thank you

**Comment by brenuart** [comment, marked as answer]
The Pattern json provider is using a standard Logback PatternLayout under the hood to resolve placeholders and converters in the pattern string. The resulting string is then converted into JSON and added to the final result. The `omitEmptyFields` property works by telling Jackson to omit fields whose value is considered empty. 

When a property does not exist, the `%property{}` converter returns `null` and unfortunately the PatternLayout appends this value "asis" to its internal StringBuilder. So in your case your pattern is converted into the string `{"container-namespace":"null"}` before it is transformed into a valid JSON structure by Jackson. As you case see, the property is not _null_ but is assigned a String value... it is therefore not empty and will not be filtered out by the `omitEmptyField` feature.

A String property is considered empty if it is _empty_ or _null_. A possible approach is to default the property to an empty string if it does not exist (see [Default value for variables](https://logback.qos.ch/manual/configuration.html#defaultValuesForVariables)). Something like this:
```
<pattern>
    <omitEmptyFields>true</omitEmptyFields>
    <pattern>
        {
            "TraceId":"%mdc{trace_id}",
            "SpanId":"%mdc{span_id}",
            "Resource": {"container-namespace":"${CONTAINER_NAMESPACE:-}"}
        }
    </pattern>
</pattern>
```

Unfortunately this won't work as expected and the variable is resolved into `CONTAINER_NAMESPACE_IS_UNDEFINED` unless the default value is set to a non empty string (See this [issue](https://jira.qos.ch/browse/LOGBACK-1012) for more information). Logback configuration does not like empty strings I'm afraid :-(

An alternative when using SpringBoot is to leveraged the `<springProperty>` custom action as follows:
```
<springProperty scope="context" name="CONTAINER_NAMESPACE" source="container.namespace" defaultValue="" />

<appender ...>
    <encoder class="net.logstash.logback.encoder.LoggingEventCompositeJsonEncoder">
        <providers>
            <pattern>
                <omitEmptyFields>true</omitEmptyFields>
                <pattern>
                    {
                        "TraceId":"%mdc{trace_id}",
                        "SpanId":"%mdc{span_id}",
                        "Resource": {"container-namespace":"${CONTAINER_NAMESPACE}"}
                    }
                </pattern>
            </pattern>
```

Finally, if your CONTAINER_NAMESPACE property is set once and for all when the application is started (i.e. does not change at runtime), then you could guard part of thee Logback configuration elements with a Janino `<if>` conditional statement so it would be included only if the property is defined.

Hope this helps.
Don't hesitate to come back with additional feedback if needed.

PS: `%mdc{}` works as expected because this converter returns an empty string instead of null when the MDC key does not exist...
PS2: I agree this is rather confusing and not very obvious... I'll see if we can do something about it in future releases...

**Reply by ziomill** [reply]
Hello @brenuart and thank you for the very exhaustive explanation.

I was just looking at the **conditional logback capabilities** provided by Janino, and I can confirm that it works as expected.
In my use cases, the environments variables are populated only once during the application startup and they never change, so I can go with Janino.

**Comment by ceki** [comment]
As a stop gap measure, you can define a variable named EMPTY with the value "" and use it as the default value:

```
<configuration>

  <variable name="EMPTY" value="" />
  ...  
    <pattern>
      <omitEmptyFields>true</omitEmptyFields>
      <pattern>
          {
            "Resource": {"container-namespace":"${CONTAINER_NAMESPACE:-${EMPTY}}"}
         }
      </pattern>
   </pattern>
   ....
</configuration>
```
